### PR TITLE
Fix settings panel fields not overwriting values from environment variables

### PR DESCRIPTION
**Pull request by jtquach1**
## Describe your changes

Please include a summary of the changes and the related issue/task. Please also include relevant motivation and context. List any dependencies that are required for this change, including links to other pull requests/branches in other repositories if applicable.

## Issue ticket number and Jira link

Please include the Jira Ticket Number and Link for this issue/task.
[REMS-496](https://jira.mitre.org/browse/REMS-496)

## Checklist before requesting a review
- [x] I have performed a self-review of my code
- [x] Ensure the target / base branch for any feature PR is set to `dev` not main (the only exception to this is releases from `dev` and hotfix branches)

## Checklist for conducting a review
- [ ] Review the code changes and make sure they all make sense and are necessary. 
- [ ] Pull the PR branch locally and test by running through workflow and making sure everything works as it is supposed to. 

## Workflow 

Owner of the Pull Request will be responsible for merge after all requirements are met, including approval from at least one reviewer. Additional changes made after a review will dismiss any approvals and require re-review of the additional updates. Auto merging can be enabled below if additional changes are likely not to be needed. The bot will auto assign reviewers to your Pull Request for you.

### CODE REVIEW

**Comment by jtquach1** [comment]
# Steps to verify the settings pane input fields, checkboxes, and buttons "work"
Note, some of these steps include details about potential "bugs".

## Input fields
1. `Base Server`
    - Click the "Select a Patient" button and select the patient "Jon Snow" with any drug.
    - Click the "Send Rx to Pharmacy" button.
    - Click the "Sign Order" button.
    - Click any of the ETASU form buttons, like "Patient Enrollment Form". 
    - How to tell it "works": On the new tab launched from clicking that button, notice the URL uses the value specified in this field. Open DevTools. If the url is invalid, you will see a 404 Not Found and an error message about an OperationOutcome not being found at that url. 
2. `Default User`
    - Click the "Select a Patient" button and select the patient "Jon Snow" with any drug. 
    - Bug?: **The practitioner name does not change under the "Prefetched" header. This is probably because the prefetch is hardcoded from the APIs located at the "Base Server" or "REMS Admin"** But clicking "Clear EHR Questionnaire Responses" will "fail" to delete QuestionnaireForms under a practitioner with a different user id for some drug. 
    - How to tell it "works": Select the patient with that drug and see that the forms created by the other user id (e.g. `pra1234`) are still there.
3. `DTR Launch URL (QuestionnaireForm)` 
    - Click the "Select a Patient" button and select the patient "Jon Snow" with any drug.
    - Click the "Send Rx to Pharmacy" button.
    - Click the "Sign Order" button.
    - Click "Patient Enrollment Form". 
    - Click "Save to EHR". 
    - Modify the field and then click the "Open In-progress Form" button. Change the url to use a different port number. 
    - How to tell it "works": If the url still uses the same port specified in your env, the form will still load "properly". If the url is invalid, you will see an error message in DevTools. 
4. `EHR Server`
    - Click the "Select a Patient" button and select the patient "Jon Snow" with any drug.
    - How to tell it "works": If you provided an invalid url, you will see an error message in DevTools.
5. `In Progress Form Expiration Days` 
    - Bug?: **There may not be a reliable way to test this**, since specifying 0 days or even -4 days brings you back any and all possible in-progress forms specified at `EHR Server`. Normally, I'd start any of the ETASU forms, click "Save to EHR", and see what list of forms are under "In Progress Form" in the patient modal.
6. `Order Select Rest End Point`
    - Bug?: **No way to test this**, as request-generator only uses `"order-sign"`.
7. `Order Sign Rest End Point`
    - Modify the field.
    - Click the "Select a Patient" button and select the patient "Jon Snow" with any drug.
    - Click the "Send Rx to Pharmacy" button.
    - Click the "Sign Order" button.
    - How to tell it "works": If you provided an invalid url, you will not see CDS Hooks cards come back and there will be an error in DevTools indicating just that.
8.  `PIMS Server`
    - Modify the field. 
    - Click the "Select a Patient" button and select the patient "Jon Snow" with any drug.
    - Click the "Send Rx to Pharmacy" button.
    - How to tell it "works": Check DevTools for an ERR_CONNECTION_REFUSED. 
9.  `REMS Admin` 
    - Modify the field. It doesn't matter what part of the original url you change.
    - Click the "Select a Patient" button and select the patient "Jon Snow" with any drug.
    - Click the "Send Rx to Pharmacy" button.
    - Click the "Sign Order" button.
    - How to tell it "works": Check DevTools for a 404 Not Found. 
10.  `SMART App`
    - Modify the field, at least the "origin" part of the url, such as `http://localhost:4040`. If you add a string after the origin, that will result in a "valid" page load. For example, use an invalid port like `64040`.
    - Click the "Select a Patient" button and select the patient "Jon Snow" with any drug.
    - Click the "Launch SMART on FHIR App" button.
    - How to tell it "works": The page will not load and if you open DevTools, there will be a 404 Not Found indicating that.

## Checkboxes
1. `Alternative Therapy Cards Allowed` - Unsure.
2. `Include Configuration in CRD Request` - Unsure.
3. `Send Prefetch` - Unsure, check the Network tab for if there's an extra request sent when the checkbox is checked?

## Buttons
1. `Clear EHR QuestionnaireResponses`
    - Modify the field. It doesn't matter what part of the original url you change.
    - How to tell it "works": You will get a 404 Not Found in DevTools if you click the button.
2. `Reset PIMS Database`
    - Modify the field, `PIMS Server`, at least the "origin" part of the url, such as `http://localhost:5051`. If you add a string after the origin, clicking the button will still "work" since the reset endpoint builds off of the origin. For example, use an invalid port like `65051`.
    - How to tell it "works": You will get a 404 Not Found in DevTools if you click the button.
3. `Reset REMS-Admin Database`
    - Modify the field, `REMS Admin`, at least the "origin" part of the url, such as `http://localhost:8090`. If you add a string after the origin, clicking the button will still "work" since the reset endpoint builds off of the origin. For example, use an invalid port like `68090`.
    - How to tell it "works": You will get a 404 Not Found in DevTools if you click the button.

**Review by KeeyanGhoreshi — CHANGES_REQUESTED** [review]

File: src/components/SettingsBox/SettingsBox.js
Comment (jtquach1) [reply src/components/SettingsBox/SettingsBox.js:23]: I reckoned the content (from RequestBuilder) should be separate from how the content is rendered (in SettingsBox). I think having the display information in the RequestBuilder clogs up RequestBuilder, as it mainly defines callbacks to be passed into its children components which do the rendering of those props. Originally, the headers were split across data.js and RequestBuilder, which looked redundant. 

I was inspired by some semblance of Model-View-Controller, so I wanted to have at least the view separated into the child. The model (?) and the controller is in the parent. 

One way I can think of the headers being in one place (`display`, `type`, and `value` keys) is by wrapping the original `value`s in lambdas, like so `() =>`. I suppose I'd still end up having to pass the state in somehow. Maybe it's cleaner than mapping?

Comment (KeeyanGhoreshi) [reply src/components/SettingsBox/SettingsBox.js:23]: SettingsBox is a stateless react component, it only renders.  All information it needs to render should be passed to it through props, not maintained inside the component.  

I don't see why the display, type, and value can't all just be in `RequestBuilder`.  In fact the whole thing should be part of the state which is both propagated to dependents and modified within the state.  

That is to say, there should be no redundancy between the state of the RequestBuilder, where it first defines something e.g. `alternativeTherapy`, and the rendering information e.g. type/display/value.  It should all just be part of the state or defined in one place, in a list or object, with no separation at all.

The only issue would be "settings" which are buttons, but those should not be with the rest of the settings anyway.  They don't modify the state, they just use it, and they break the pattern of callback to state update which is confusing.  They should be pulled out and put into the settings box along with their associated functions, or they could be separated out into another prop that gets passed to SettingsBox to be rendered, separate from the actual settings.

Comment (jtquach1) [reply src/components/SettingsBox/SettingsBox.js:23]: I agree with this part of your explanation especially (not to say I do not agree with the text before it, just drawing out this part):
> It should all just be ... defined in one place, in a list or object, with no separation at all. 

I don't think the `display` and `type` should be in the state since there's never a way for the user to modify the input/checkbox headers or how the content should be rendered (as an input or a checkbox). But I think all the React.js components (like SettingsBox) should fall under the category of a "controller" and the raw, changeable strings that are the `display` could be moved to some place like data.js. React.js sort of blurs the lines of MVC anyways... 

> The only issue would be "settings" which are buttons, but those should not be with the rest of the settings anyway. They don't modify the state, they just use it, and they break the pattern of callback to state update which is confusing. They should be pulled out and put into the settings box along with their associated functions, or they could be separated out into another prop that gets passed to SettingsBox to be rendered, separate from the actual settings.
I agree. I've not yet brainstormed an approach to this- to me the reset buttons are part of the "kitchen utilities/junk drawer" of the SettingsBox component. @plarocque4 what are your thoughts?

File: src/util/data.js
Comment (KeeyanGhoreshi) [inline src/util/data.js:91]: this concerns me a bit, given that we're kind of limiting ourselves when it comes to functions we can put in the settings.  I'd rather just take the button type out of this abstraction entirely and do those manually inside the settings box.  There's really no need to have them in a list like this since they don't interact with the state.

File: src/util/data.js
Comment (KeeyanGhoreshi) [inline src/util/data.js:75]: these definitions are a little disjointed.  They are passed to SettingsBox, which use them to associate a key with a state value in RequestBox, but the state value in RequestBox is not based on these definitions.  

We'd probably be better off getting rid of this definition and making a `settings` object in the RequestBox state that we update.

File: src/util/data.js
Comment (jtquach1) [reply src/util/data.js]: Done as of https://github.com/mcode/request-generator/pull/53/commits/06f848b1febbf1c6dc7c661bbaaafdfe3a4f5fff

File: src/util/data.js
Comment (jtquach1) [reply src/util/data.js:1]: I think I am lost on your comment, in terms of specific code changes. RequestBox uses the state from the parent component RequestBuilder, which is then passed into and mutated by SettingsBox (so RequestBox and SettingsBox are siblings).

Comment (jtquach1) [reply src/util/data.js:1]: I do notice that despite changing the "default user", I'm still seeing pra1234 in the prefetch (from RequestBox, which uses the `submit_info` function from RequestBuilder which does have a `this.state.defaultUser`.

Comment (jtquach1) [reply src/util/data.js:1]: image

Comment (KeeyanGhoreshi) [reply src/util/data.js:1]: Sorry, I said RequestBox when I meant RequestBuilder.  The basic idea here is that the name of the key is being defined in two different places.  This should be unnecessary.  SettingsBox could potentially make an input that updates a state that doesn't actually exist and it would produce no error.  

In terms of code changes... this would basically require an overhaul of the whole system, so that we have only one "source of truth" which defines the RequestBox's state AND the inputs that get automatically generated.  It's probably not worth it.

File: src/components/Inputs/CheckBox.js
Comment (KeeyanGhoreshi) [inline src/components/Inputs/CheckBox.js:14]: this no longer updates the state, makes checkboxes useless

File: src/components/SettingsBox/SettingsBox.js
Comment (KeeyanGhoreshi) [reply src/components/SettingsBox/SettingsBox.js]: sure, don't put the display and the type in the state.  It's not particularly important.  If the settings are to be derived from some JSON object, then the state should be derived from the same object.  There should be no ambiguity between the state and the definition about what the settings are named.  

If you have a JSON definition of name/type/initialValue/display, you can generate the state of the settingsBox from that JSON, using the name and initialValue, and then pass the JSON itself into settingsBox.  Now the state and the settings are synced.  

Right now we are halfway between a dynamic and static approach.

Comment (KeeyanGhoreshi) [reply src/components/SettingsBox/SettingsBox.js]: sure, don't put the display and the type in the state. It's not particularly important. If the settings are to be derived from some JSON object, then the state should be derived from the same object. There should be no ambiguity between the state and the definition about what the settings are named.

If you have a JSON definition of name/type/initialValue/display, you can generate the state of the settingsBox from that JSON, using the name and initialValue, and then pass the JSON itself into settingsBox. Now the state and the settings are synced.

Right now we are halfway between a dynamic and static approach.

File: src/util/data.js
Comment (jtquach1) [reply src/util/data.js:1]: Maybe we can create a ticket/sub-ticket (?) as part of refactoring/making tests for request generator (would that be an epic? I am unsure of the right terminology).

File: src/components/Inputs/CheckBox.js
Comment (jtquach1) [reply src/components/Inputs/CheckBox.js]: Whoops, I missed that. Added back as of [721b6b2](https://github.com/mcode/request-generator/pull/53/commits/721b6b2e25e78cb8453d74ce25336a9c83d243dc).

File: src/components/Inputs/CheckBox.js
Comment (KeeyanGhoreshi) [inline src/components/Inputs/CheckBox.js:16]: setting the state then accessing the state here might cause an issue.  Pull out the `!this.state.toggle` as a variable to avoid uncertainty.

File: src/util/data.js
Comment (KeeyanGhoreshi) [reply src/util/data.js:1]: make a low priority ticket for refactoring the settings, call it a day

File: src/components/Inputs/CheckBox.js
Comment (jtquach1) [reply src/components/Inputs/CheckBox.js]: I resolved this potential uncertainty via https://github.com/mcode/request-generator/pull/53/commits/10f2574b83f3a58ef64e289a9db7b909a6f69800, although through just removing the internal state completely. I'm unaware of whether that will cause any performance issues, but it seems like they work fine. 

e.g. I just tried
1. changing the PIMS url (InputBox) then clicking on the "Reset PIMS" button to see the state gets passed down to and mutated by the InputBox component properly. 
2. selecting a patient and drug, clicking the "Send Rx to Pharmacy" button, then clicking the "Sign Order" button to see the rems-order-sign Request is sent in the Network tab. Then clicking the "Send Prefetch" checkbox to see whether that rems-order-sign shows up (it doesn't, good). Although unfortunately, re-checking the checkbox afterwards doesn't re-establish the connection to 127.0.0.1:8080, at least while using Docker (I'm unaware what the behavior is like just running our tech stack without Docker).

File: src/util/data.js
Comment (jtquach1) [reply src/util/data.js:1]: Done, the ticket is https://jira.mitre.org/browse/REMS-540

**Review by KeeyanGhoreshi — APPROVED** [review]

File: src/components/SettingsBox/SettingsBox.js
Comment (jtquach1) [reply src/components/SettingsBox/SettingsBox.js]: I... missed this comment, sorry! 

Your comment reminds me of the latest React API (with the functional components) and passing an initial state in and using a reducer. Maybe I can try doing that in that request-generator refactoring ticket one day.

**Comment by jtquach1** [comment]
Just clicked the red merge button since I found a typo and pushed that tiny change after merging in the changes from origin/dev- but that made the review stale. The newer commits had merge changes for imports and said typo fix, so I don't think that warrants another review. Sorry for clicking the request review again @KeeyanGhoreshi, that was before I figured out how to force the merge.

**Pull request merged**
